### PR TITLE
Parametrize validate_on_assignment tests

### PR DESCRIPTION
**Pull request by jdavies-st**
Hi @kmacdonald-stsci, this is just a clean-up of the tests by parametrizing them.  Hopefully in the parameters, you can see the expected behavior a bit more clearly?

I added a context manager called `does_not_raise` that works like `pytest.raises` but allows for no errors.  So it works like

```python
with pytest.raises(SomeError):
    pass
```

but like

```python
with does_not_raise:
    pass
```

but does not actually check any errors.  So if an error was actually raised in the context, it would fail the test.  It allows for parametrization and keeping the test code clear.  It allowed cleanup of `test_strict_validation_attribute_assignment()` test so that we don't have the `issubclass()` stuff in there.

https://github.com/kmacdonald-stsci/stdatamodels/pull/1/files#diff-f1dbe79adfb6e39e76ee4519368bc4b5eaa0e4831f00165e5ac28fede4825bffL161-L169

Same on your tests.

I also removed the repeated testing of the same things that were tested in other tests, as it simplified what is being tested in each.

If this is good, feel free to merge, and then I think we can merge yours into `stdatamodels` master.  If there's any issues, let's hash them out in this PR.

### CODE REVIEW

**Comment by jdavies-st** [comment]
Ah, it looks like CI is not set up to run on your fork.  Here's what I see locally:

```
> pytest tests/test_validation.py
=============================== test session starts ================================
platform darwin -- Python 3.9.0, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /Users/jdavies/dev/stdatamodels, configfile: setup.cfg
plugins: doctestplus-0.8.0, asdf-2.8.0.dev107+g8ccee8b, cov-2.10.1, openfiles-0.5.0
collected 38 items                                                                 

tests/test_validation.py ...xx......x......x.x.................              [100%]

========================== 33 passed, 5 xfailed in 1.43s ===========================
```

File: tests/test_validation.py [hunk 1]
```diff
@@ -1,13 +1,31 @@
...
 import asdf
 from jsonschema import ValidationError
 
 from stdatamodels.validate import ValidationWarning
-
 from models import ValidationModel, RequiredModel
 
 
+
+class _DoesNotRaiseContext:
```
Comment (eslavich) [inline tests/test_validation.py:12]: Cool!  This is clever, I like it.

**Review by eslavich — APPROVED** [review]
👍

File: tests/test_validation.py (hunk 1 above)
Comment (jdavies-st) [reply tests/test_validation.py:12]: There was actually a PR that the maintainers of `pytest` had that was going to implement this feature a few years ago, and this is the implementation.  But they couldn't agree on an interface for it, so closed the PR and never merged it.  That doesn't mean we can't use it.  =)

**Comment by kmacdonald-stsci** [comment]
Forget big endian for the moment.  Create a new branch.  Use the github action and create a new pull request to make sure everything works.

**Comment by kmacdonald-stsci** [comment]
https://learning.oreilly.com/videos/elegant-automation-frameworks/9781800561489/

**Comment by jdavies-st** [comment]
@kmacdonald-stsci  you seem to be sending emails to this PR.

**Comment by kmacdonald-stsci** [comment]
Moved subarray handling to the step, which means changing the tests, so that gain and read noise are created as arrays and not models.

**Pull request merged**